<h2 align='center' style='color:blue'>Dropout Regularization In Deep Neural Network</h2>

This is a dataset that describes sonar chirp returns bouncing off different services. The 60 input variables are the strength of the returns at different angles. It is a **binary classification problem** that requires a model to differentiate rocks from metal cylinders.

Dataset information: https://archive.ics.uci.edu/ml/datasets/Connectionist+Bench+(Sonar,+Mines+vs.+Rocks)
Download it from here: https://archive.ics.uci.edu/ml/machine-learning-databases/undocumented/connectionist-bench/sonar/sonar.all-data

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
import warnings
warnings.filterwarnings('ignore')

In [3]:
df = pd.read_csv("./sonar_dataset.csv", header=None)
df.sample(5)

0       1       2       3       4       5       6       7       8   \
89   0.0235  0.0291  0.0749  0.0519  0.0227  0.0834  0.0677  0.2002  0.2876   
80   0.0229  0.0369  0.0040  0.0375  0.0455  0.1452  0.2211  0.1188  0.0750   
60   0.0130  0.0006  0.0088  0.0456  0.0525  0.0778  0.0931  0.0941  0.1711   
170  0.0134  0.0172  0.0178  0.0363  0.0444  0.0744  0.0800  0.0456  0.0368   
101  0.0335  0.0134  0.0696  0.1180  0.0348  0.1180  0.1948  0.1607  0.3036   

         9   ...      51      52      53      54      55      56      57  \
89   0.3674  ...  0.0083  0.0037  0.0095  0.0105  0.0030  0.0132  0.0068   
80   0.1631  ...  0.0064  0.0022  0.0122  0.0151  0.0056  0.0026  0.0029   
60   0.1483  ...  0.0078  0.0041  0.0013  0.0011  0.0045  0.0039  0.0022   
170  0.1250  ...  0.0094  0.0048  0.0047  0.0016  0.0008  0.0042  0.0024   
101  0.4372  ...  0.0244  0.0232  0.0093  0.0159  0.0193  0.0032  0.0377   

         58      59  60  
89   0.0108  0.0090   R  
80   0.0104  0.0163   R  
60   0.0023  0.0016   R  
170  0.0027  0.0041   M  
101  0.0126  0.0156   M  

[5 rows x 61 columns]

In [4]:
df.shape

(208, 61)

In [5]:
# check for nan values
df.isna().sum()

0     0
1     0
2     0
3     0
4     0
     ..
56    0
57    0
58    0
59    0
60    0
Length: 61, dtype: int64

In [6]:
df.columns

Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35,
       36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53,
       54, 55, 56, 57, 58, 59, 60],
      dtype='int64')

In [7]:
df[60].value_counts() # label is not skewed

60
M    111
R     97
Name: count, dtype: int64

In [8]:
X = df.drop(60, axis=1)
y = df[60]
y.head()

0    R
1    R
2    R
3    R
4    R
Name: 60, dtype: object

In [9]:
y = pd.get_dummies(y, drop_first=True)
y.sample(5) # R --> 1 and M --> 0

R
59    True
153  False
71    True
92    True
52    True

In [10]:
y.value_counts()

R    
False    111
True      97
Name: count, dtype: int64

In [11]:
X.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      50      51      52      53      54      55      56  \
0  0.2111  ...  0.0232  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180   
1  0.2872  ...  0.0125  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140   
2  0.6194  ...  0.0033  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316   
3  0.1264  ...  0.0241  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050   
4  0.4459  ...  0.0156  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072   

       57      58      59  
0  0.0084  0.0090  0.0032  
1  0.0049  0.0052  0.0044  
2  0.0164  0.0095  0.0078  
3  0.0044  0.0040  0.0117  
4  0.0048  0.0107  0.0094  

[5 rows x 60 columns]

In [12]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

In [13]:
X_train.head()

0       1       2       3       4       5       6       7       8   \
67   0.0368  0.0403  0.0317  0.0293  0.0820  0.1342  0.1161  0.0663  0.0155   
14   0.0124  0.0433  0.0604  0.0449  0.0597  0.0355  0.0531  0.0343  0.1052   
164  0.0163  0.0198  0.0202  0.0386  0.0752  0.1444  0.1487  0.1484  0.2442   
179  0.0394  0.0420  0.0446  0.0551  0.0597  0.1416  0.0956  0.0802  0.1618   
19   0.0126  0.0149  0.0641  0.1732  0.2565  0.2559  0.2947  0.4110  0.4983   

         9   ...      50      51      52      53      54      55      56  \
67   0.0506  ...  0.0058  0.0091  0.0160  0.0160  0.0081  0.0070  0.0135   
14   0.2120  ...  0.0078  0.0083  0.0057  0.0174  0.0188  0.0054  0.0114   
164  0.2822  ...  0.0027  0.0077  0.0026  0.0031  0.0083  0.0020  0.0084   
179  0.2558  ...  0.0118  0.0146  0.0040  0.0114  0.0032  0.0062  0.0101   
19   0.5920  ...  0.0153  0.0092  0.0035  0.0098  0.0121  0.0006  0.0181   

         57      58      59  
67   0.0067  0.0078  0.0068  
14   0.0196  0.0147  0.0062  
164  0.0108  0.0083  0.0033  
179  0.0068  0.0053  0.0087  
19   0.0094  0.0116  0.0063  

[5 rows x 60 columns]

# Using Deep Learning Model

## Model without Dropout Layer

In [14]:
import tensorflow as tf
from tensorflow import keras

In [15]:
model = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid')
])

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

model.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 561us/step - accuracy: 0.4605 - loss: 0.7025
Epoch 2/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 488us/step - accuracy: 0.6221 - loss: 0.6597
Epoch 3/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 434us/step - accuracy: 0.6864 - loss: 0.6370
Epoch 4/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 550us/step - accuracy: 0.6970 - loss: 0.5983
Epoch 5/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 483us/step - accuracy: 0.7755 - loss: 0.5527
Epoch 6/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 440us/step - accuracy: 0.8079 - loss: 0.5250
Epoch 7/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 416us/step - accuracy: 0.6936 - loss: 0.5449
Epoch 8/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 439us/step - accuracy: 0.7904 - loss: 0.4410
Epoch 9/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 430us/step - accuracy: 0.8277 - loss: 0.4362
Epoch 10/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 434us/step - accuracy: 0.8824 - loss: 0.3789
Epoch 11/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 430us/step - accuracy: 0.8843 - loss: 0.3427
Epoch 12/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0

20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 391us/step - accuracy: 1.0000 - loss: 0.0070
Epoch 69/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 402us/step - accuracy: 1.0000 - loss: 0.0064
Epoch 70/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 399us/step - accuracy: 1.0000 - loss: 0.0080
Epoch 71/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 393us/step - accuracy: 1.0000 - loss: 0.0064
Epoch 72/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 396us/step - accuracy: 1.0000 - loss: 0.0050
Epoch 73/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 396us/step - accuracy: 1.0000 - loss: 0.0051
Epoch 74/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 391us/step - accuracy: 1.0000 - loss: 0.0051
Epoch 75/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 404us/step - accuracy: 1.0000 - loss: 0.0041
Epoch 76/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 401us/step - accuracy: 1.0000 - loss: 0.0082
Epoch 77/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 408us/step - accuracy: 1.0000 - loss: 0.0050
Epoch 78/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 399us/step - accuracy: 1.0000 - loss: 0.0046
Epoch 79/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 39

In [16]:
model.evaluate(X_test, y_test)

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.7163 - loss: 1.2283 


[1.1178494691848755, 0.7307692170143127]

Training Accuracy >>> Test Accuracy

In [17]:
y_pred = model.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 15ms/step
[2.2697327e-07 9.8939717e-01 9.9741375e-01 2.1143735e-05 9.9999911e-01
 9.9980646e-01 5.4988261e-02 9.9999636e-01 7.5304597e-06 9.9999970e-01]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [18]:
y_test[:10]

R
186  False
155  False
165  False
200  False
58    True
34    True
151  False
18    True
202  False
62    True

In [19]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

       False       0.70      0.85      0.77        27
        True       0.79      0.60      0.68        25

    accuracy                           0.73        52
   macro avg       0.74      0.73      0.72        52
weighted avg       0.74      0.73      0.73        52



### Model with Dropout Layer

In [20]:
modeld = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(1, activation='sigmoid')
])

modeld.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

modeld.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 503us/step - accuracy: 0.5264 - loss: 0.7238
Epoch 2/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 514us/step - accuracy: 0.5469 - loss: 0.6855
Epoch 3/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 520us/step - accuracy: 0.5933 - loss: 0.6811
Epoch 4/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 524us/step - accuracy: 0.5946 - loss: 0.6933
Epoch 5/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 458us/step - accuracy: 0.5124 - loss: 0.7228
Epoch 6/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 435us/step - accuracy: 0.5894 - loss: 0.6667
Epoch 7/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 428us/step - accuracy: 0.5347 - loss: 0.6818
Epoch 8/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 444us/step - accuracy: 0.5340 - loss: 0.7118
Epoch 9/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 479us/step - accuracy: 0.5374 - loss: 0.6964
Epoch 10/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 457us/step - accuracy: 0.5318 - loss: 0.6821
Epoch 11/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 501us/step - accuracy: 0.5401 - loss: 0.6662
Epoch 12/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0

20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 436us/step - accuracy: 0.8766 - loss: 0.3333
Epoch 69/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 436us/step - accuracy: 0.8519 - loss: 0.4221
Epoch 70/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 452us/step - accuracy: 0.8268 - loss: 0.3702
Epoch 71/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 441us/step - accuracy: 0.8412 - loss: 0.4362
Epoch 72/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 449us/step - accuracy: 0.8889 - loss: 0.3246
Epoch 73/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 463us/step - accuracy: 0.8188 - loss: 0.3787
Epoch 74/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 435us/step - accuracy: 0.8489 - loss: 0.3401
Epoch 75/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 455us/step - accuracy: 0.8206 - loss: 0.4043
Epoch 76/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 455us/step - accuracy: 0.8534 - loss: 0.3520
Epoch 77/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 447us/step - accuracy: 0.8343 - loss: 0.4696
Epoch 78/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 454us/step - accuracy: 0.7816 - loss: 0.3997
Epoch 79/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 44

In [21]:
modeld.evaluate(X_test, y_test)

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7524 - loss: 0.5126 


[0.47256845235824585, 0.7692307829856873]

Training Accuracy is still good but Test Accuracy Improved

In [22]:
y_pred = modeld.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 19ms/step
[8.3189993e-04 7.9852378e-01 9.4774294e-01 1.2075177e-02 9.9979472e-01
 9.4851738e-01 3.2824168e-01 9.9976289e-01 1.8211395e-02 9.9988556e-01]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [23]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

       False       0.74      0.85      0.79        27
        True       0.81      0.68      0.74        25

    accuracy                           0.77        52
   macro avg       0.78      0.77      0.77        52
weighted avg       0.77      0.77      0.77        52



**You can see that by using dropout layer test accuracy increased from 0.73 to 0.77**